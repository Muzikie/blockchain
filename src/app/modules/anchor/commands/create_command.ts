/* eslint-disable @typescript-eslint/member-ordering */
/* eslint-disable class-methods-use-this */
import {
  BaseCommand,
  CommandVerifyContext,
  CommandExecuteContext,
  VerificationResult,
  VerifyStatus,
} from 'lisk-sdk';
import { AnchorStore } from '../stores/anchor';
import { AnchorAccountStore } from '../stores/anchorAccount';
import { AnchorCreated } from '../events/anchorCreated';
import { CreateCommandParams, Anchor, AnchorAccount } from '../types';
import { createCommandParamsSchema } from '../schemas';
import { getCreatedAt, getAnchorID } from '../utils';
import { VOTE_RATE_LIMIT } from '../constants';
import { BadgeMethod } from '../../badge/method';
import { Badges } from '../../badge/types';
import { getBadgeID } from '../../badge/utils';
import { AnchorStatsStore } from '../stores/anchorStats';

export class CreateCommand extends BaseCommand {
  public schema = createCommandParamsSchema;
  private _badgeMethod!: BadgeMethod;

  public addDependencies(badgeMethod: BadgeMethod) {
    this._badgeMethod = badgeMethod;
  }

  // eslint-disable-next-line @typescript-eslint/require-await
  public async verify(
    context: CommandVerifyContext<CreateCommandParams>,
  ): Promise<VerificationResult> {
    const anchorStore = this.stores.get(AnchorStore);
    const { senderAddress } = context.transaction;
    const anchorAccountStore = this.stores.get(AnchorAccountStore);
    const anchorId = getAnchorID(context.params);

    const anchorExists = await anchorStore.has(context, anchorId);
    if (anchorExists) {
      throw new Error('This anchor already exist.');
    }

    // Add submission rate limit
    const senderExists = await anchorAccountStore.has(context, senderAddress);
    if(senderExists) {
      const senderAccount = await anchorAccountStore.get(context, senderAddress);
      const IDS = senderAccount.anchors.slice(-1 * VOTE_RATE_LIMIT);

      if(IDS.length >= VOTE_RATE_LIMIT) {
        const anchor = await anchorStore.get(context, IDS[0]);

        if (anchor.createdAt === getCreatedAt(Math.floor(((new Date()).getTime())))) {
          throw new Error(`You have exceeded the ${VOTE_RATE_LIMIT} anchor submissions daily limit.`);
        }
      }
    }

    return { status: VerifyStatus.OK };
  }

  public async execute(context: CommandExecuteContext<CreateCommandParams>): Promise<void> {
    const {
      params,
      transaction: { senderAddress },
    } = context;
    const anchorAccountStore = this.stores.get(AnchorAccountStore);
    const anchorStore = this.stores.get(AnchorStore);
    const anchorStatsStore = this.stores.get(AnchorStatsStore);
    const methodContext = context.getMethodContext();

    // Create anchor ID
    const anchorID = getAnchorID(context.params);
    const createdAt = getCreatedAt(new Date().getTime());
    // Create anchor object
    const anchor: Anchor = {
      ...params,
      id: anchorID,
      createdAt,
      votes: [],
      submitter: senderAddress,
    };

    const badgeIDs = [1, 2, 3]
      .map((rank) => getBadgeID(createdAt, rank, Badges.AOTD));
    context.logger.info(badgeIDs.map((idd) => idd.toString('hex')));
    // Store the anchor object in the blockchain
    await anchorStore.set(context, anchorID, anchor);

    const StatsExist = await anchorStatsStore.has(context, Buffer.from(createdAt));
  
    if (!StatsExist) {
      const newAnchorStats = {
        date: createdAt,
        anchorsCount: 1,
        votesCount: 0,
      };
      await anchorStatsStore.set(context, Buffer.from(createdAt), newAnchorStats);
      context.logger.info(`Anchors count for ${createdAt}: 1`);
    } else {
      const anchorStats = await anchorStatsStore.get(context, Buffer.from(createdAt));
      anchorStats.anchorsCount += 1;
      await anchorStatsStore.set(context, Buffer.from(createdAt), anchorStats);
      context.logger.info(`Anchors count2SS for ${createdAt}: ${anchorStats.anchorsCount}`);
    }

    // Get the sender account from the blockchain
    const senderExists = await anchorAccountStore.has(context, senderAddress);
    let senderAccount: AnchorAccount;
    if (!senderExists) {
      senderAccount = {
        anchors: [anchorID],
        votes: [],
      };
    } else {
      const retrievedAccount = await anchorAccountStore.get(context, senderAddress);
      senderAccount = {
        anchors: [...retrievedAccount.anchors, anchorID],
        votes: retrievedAccount.votes
      };
    }

    // Store the account object in the blockchain
    await anchorAccountStore.set(context, senderAddress, senderAccount);

    await this._badgeMethod.createBadgesForDay(methodContext, createdAt);

    const anchorCreated = this.events.get(AnchorCreated);
    anchorCreated.add(context, {
      submitter: context.transaction.senderAddress,
      anchorID,
      createdAt,
    }, [context.transaction.senderAddress]);
  }
}
