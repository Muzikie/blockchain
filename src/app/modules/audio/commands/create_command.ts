/* eslint-disable class-methods-use-this */
import {
  BaseCommand,
  CommandVerifyContext,
  CommandExecuteContext,
  VerificationResult,
  VerifyStatus,
} from 'lisk-sdk';
import { AudioStore } from '../stores/audio';
import { AudioAccountStore } from '../stores/audioAccount';
import { CreateCommandParams, Audio, AudioAccount } from '../types';
import { createCommandParamsSchema } from '../schemas';
import { VALID_GENRES, MIN_RELEASE_YEAR } from '../constants';
import { getEntityID, verifyHash } from '../../../utils';
import { AudioCreated } from '../events/audioCreated'

export class CreateCommand extends BaseCommand {
  public schema = createCommandParamsSchema;

  // eslint-disable-next-line @typescript-eslint/require-await
  public async verify(
    context: CommandVerifyContext<CreateCommandParams>,
  ): Promise<VerificationResult> {
    const { params, transaction } = context;
    const thisYear = new Date().getFullYear();
    const numericYear = Number(params.releaseYear);
    if (numericYear < MIN_RELEASE_YEAR || numericYear > thisYear) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error(
          `Release year must be a number between ${MIN_RELEASE_YEAR} and ${thisYear}`,
        ),
      };
    }
    if (params.genre.some(item => item > VALID_GENRES.length)) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error('Genres should be selected from the list of valid genres'),
      };
    }
    if (
      params.owners.length === 0 ||
      params.owners.some(item => item.shares < 1) ||
      params.owners.reduce((acc, item) => acc + item.shares, 0) !== 100
    ) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error(
          'Owners should have a total of 100 shares. Each owner should have a positive number of shares.',
        ),
      };
    }

    const isHashGenuine = verifyHash(params.audioSignature, params.audioHash, transaction.senderPublicKey);
    if (!isHashGenuine) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error('The hash signature is not generated by the sender'),
      };
    }

    return { status: VerifyStatus.OK };
  }

  public async execute(context: CommandExecuteContext<CreateCommandParams>): Promise<void> {
    const { params, transaction } = context;
    // Get namehash output of the audio file
    const audioID = getEntityID(context.transaction);

    const audioAccountSubStore = this.stores.get(AudioAccountStore);
    const audioSubStore = this.stores.get(AudioStore);

    // Check uniqueness of the NFT
    const audioExists = await audioSubStore.has(context, audioID);
    if (audioExists) {
      throw new Error('You have already created this audio.');
    }

    // Create the Audio object and save it on the blockchain
    const owners = params.owners.map(owner => ({
      address: owner.address,
      shares: owner.shares,
      income: BigInt(0),
    }));

    const audioObject: Audio = {
      ...params,
      owners,
      creatorAddress: transaction.senderAddress,
    };

    // Store the hash of the audio object in the sender account
    const accountExists = await audioAccountSubStore.has(context, transaction.senderAddress);
    if (accountExists) {
      const senderAccount: AudioAccount = await audioAccountSubStore.get(
        context,
        transaction.senderAddress,
      );
      senderAccount.audio.audios = [...senderAccount.audio.audios, audioID];
      await audioAccountSubStore.set(context, transaction.senderAddress, senderAccount);
    } else {
      await audioAccountSubStore.set(context, context.transaction.senderAddress, {
        audio: {
          audios: [audioID],
        },
      });
    }

    // @todo Here we should check if the Audio is already uploaded using steganography methods
    // Store the audio object in the blockchain
    await audioSubStore.set(context, audioID, audioObject);

    // Emit a "New collection" event
    const audioCreated = this.events.get(AudioCreated);
    audioCreated.add(context, {
      creatorAddress: context.transaction.senderAddress,
      audioID,
    }, [context.transaction.senderAddress]);
  }
}
