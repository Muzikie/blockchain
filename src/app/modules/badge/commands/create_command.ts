/* eslint-disable @typescript-eslint/member-ordering */
/* eslint-disable class-methods-use-this */
import {
  BaseCommand,
  CommandVerifyContext,
  CommandExecuteContext,
  VerificationResult,
  VerifyStatus,
} from 'lisk-sdk';
import { BadgeStore } from '../stores/badge';
import { BadgeAccountStore } from '../stores/badgeAccount';
import { CreateCommandParams, Badge, BadgeAccount } from '../types';
import { createCommandParamsSchema } from '../schemas';
import { TREASURY_ADDRESS } from '../../../constants';
import { BadgeCreated } from '../events/badgeCreated';

const getBadgeID = (params: CreateCommandParams) =>
  Buffer.concat([
    Buffer.from(params.type, 'utf8'),
    Buffer.from(params.rank.toString(), 'utf8'),
    Buffer.from(params.awardDate, 'utf8')],
  );

export class CreateCommand extends BaseCommand {
  public schema = createCommandParamsSchema;

  // eslint-disable-next-line @typescript-eslint/require-await
  public async verify(
    context: CommandVerifyContext<CreateCommandParams>,
  ): Promise<VerificationResult> {
    const { transaction, params } = context;

    const badgeSubStore = this.stores.get(BadgeStore);

    if (!transaction.senderAddress.equals(TREASURY_ADDRESS)) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error('You are not authorized to create a anchor.'),
      };
    }

    // Ensure that there's only one badge (of each type) per day
    const badgeID = getBadgeID(params);
    const badgeExists = await badgeSubStore.has(context, badgeID);
    if (badgeExists) {
      throw new Error('You have already created this badge.');
    }

    return { status: VerifyStatus.OK };
  }

  public async execute(context: CommandExecuteContext<CreateCommandParams>): Promise<void> {
    const { params, transaction } = context;
    // Get namehash output of the badge file
    const badgeID = getBadgeID(params);

    const badgeAccountSubStore = this.stores.get(BadgeAccountStore);
    const badgeSubStore = this.stores.get(BadgeStore);

    // Create Badge NFT
    const badgeNFT: Badge = {
      ...params,
      badgeID,
      prize: BigInt(0), // @todo calculate prize for the rank
      claimed: false,
    };

    // Store the hash of the badge object in the sender account
    const accountExists = await badgeAccountSubStore.has(context, transaction.senderAddress);
    if (accountExists) {
      const senderAccount: BadgeAccount = await badgeAccountSubStore.get(
        context,
        transaction.senderAddress,
      );
      senderAccount.badges = [...senderAccount.badges, badgeID];
      await badgeAccountSubStore.set(context, transaction.senderAddress, senderAccount);
    } else {
      await badgeAccountSubStore.set(context, context.transaction.senderAddress, {
        badges: [badgeID],
      });
    }

    // Store the badge object in the blockchain
    await badgeSubStore.set(context, badgeID, badgeNFT);

    // Emit a "New collection" event
    const badgeCreated = this.events.get(BadgeCreated);
    badgeCreated.add(context, {
      badgeID,
      prize: BigInt(0),
    }, [context.transaction.senderAddress]);
  }
}
