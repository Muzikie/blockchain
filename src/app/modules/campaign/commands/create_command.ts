/* eslint-disable class-methods-use-this */
import { Modules, StateMachine } from 'klayr-sdk';
import { CampaignStore } from '../stores/campaign';
import { CampaignAccountStore } from '../stores/campaign_account';
import { CampaignCreated } from '../events/campaign_created';
import { createCommandParamsSchema } from '../schemas';
import { CreateCommandParams, Campaign, CampaignStatus, CampaignAccount } from '../types';
import { getCampaignId } from '../utils';

export class CreateCommand extends Modules.BaseCommand {
	public async verify(
		context: StateMachine.CommandVerifyContext<CreateCommandParams>,
	): Promise<StateMachine.VerificationResult> {
		const { params } = context;
		const campaignStore = this.stores.get(CampaignStore);
		const campaignId = getCampaignId(params);

		const campaignExists = await campaignStore.has(context, campaignId);
		if (campaignExists) {
			throw new Error('This campaign already exist.');
		}

		return { status: StateMachine.VerifyStatus.OK };
	}

	public async execute(
		context: StateMachine.CommandExecuteContext<CreateCommandParams>,
	): Promise<void> {
		const {
			params,
			transaction: { senderAddress },
		} = context;

		const campaignAccountStore = this.stores.get(CampaignAccountStore);
		const campaignStore = this.stores.get(CampaignStore);

		// Create campaign ID
		const campaignId = getCampaignId(params);

		// Create campaign object
		const campaign: Campaign = {
			softGoal: BigInt(params.softGoal),
			hardGoal: BigInt(params.hardGoal),
			currentFunding: BigInt(0),
			status: CampaignStatus.Draft,
			deadline: params.deadline,
			apiId: params.apiId,
			contributionTiers: [],
			id: campaignId,
			submitter: senderAddress,
		};

		// Store the campaign object in the blockchain
		await campaignStore.set(context, campaignId, campaign);

		// Get the sender account from the blockchain
		const senderExists = await campaignAccountStore.has(context, senderAddress);
		let senderAccount: CampaignAccount;
		if (!senderExists) {
			senderAccount = {
				campaigns: [campaignId],
			};
		} else {
			const retrievedAccount = await campaignAccountStore.get(context, senderAddress);
			senderAccount = {
				campaigns: [...retrievedAccount.campaigns, campaignId],
			};
		}

		// Store the account object in the blockchain
		await campaignAccountStore.set(context, senderAddress, senderAccount);

		const campaignCreated = this.events.get(CampaignCreated);
		campaignCreated.add(
			context,
			{
				submitter: senderAddress,
				campaignId,
			},
			[senderAddress],
		);
	}

	public schema = createCommandParamsSchema;
}
