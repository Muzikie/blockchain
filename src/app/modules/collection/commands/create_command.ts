/* eslint-disable class-methods-use-this */
import {
  BaseCommand,
  CommandVerifyContext,
  CommandExecuteContext,
  VerificationResult,
  VerifyStatus,
} from 'lisk-sdk';
import { CollectionStore } from '../stores/collection';
import { CollectionAccountStore } from '../stores/collectionAccount';
import { CreateCommandParams, Collection, CollectionAccount } from '../types';
import { createCommandParamsSchema } from '../schemas';
import { validCollectionTypes, MIN_RELEASE_YEAR } from '../constants';
import { getNodeForName, verifyHash } from '../utils';

export class CreateCommand extends BaseCommand {
  public schema = createCommandParamsSchema;

  // eslint-disable-next-line @typescript-eslint/require-await
  public async verify(
    context: CommandVerifyContext<CreateCommandParams>,
  ): Promise<VerificationResult> {
    const { params, transaction } = context;

    const thisYear = new Date().getFullYear();
    const numericYear = Number(params.releaseYear);
    if (numericYear < MIN_RELEASE_YEAR || numericYear > thisYear) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error(
          `Release year must be a number between ${MIN_RELEASE_YEAR} and ${thisYear}`,
        ),
      };
    }
    if (!validCollectionTypes.includes(params.collectionType)) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error('Collection type should be selected from the list of valid types'),
      };
    }
    const isHashGenuine = verifyHash(params.hash, params.meta, transaction.senderPublicKey);
    if (!isHashGenuine) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error('The hash is not generated by the sender'),
      };
    }
    return { status: VerifyStatus.OK };
  }

  public async execute(context: CommandExecuteContext<CreateCommandParams>): Promise<void> {
    const { params, transaction } = context;
    // Get namehash output of the audio file
    const key = getNodeForName(params);

    const collectionAccountSubStore = this.stores.get(CollectionAccountStore);
    const collectionSubStore = this.stores.get(CollectionStore);

    // Check uniqueness of the NFT
    const collectionExists = await collectionSubStore.has(context, key);
    if (collectionExists) {
      throw new Error('You have already created this audio.');
    }

    // Create the Collection object and save it on the blockchain
    // Note: The audios list is initially empty
    const audioObject: Collection = {
      ...params,
      audios: [],
      creatorAddress: transaction.senderAddress,
    };

    // Store the hash of the audio object in the sender account
    const accountExists = await collectionAccountSubStore.has(context, transaction.senderAddress);
    if (accountExists) {
      const senderAccount: CollectionAccount = await collectionAccountSubStore.get(
        context,
        transaction.senderAddress,
      );
      senderAccount.collection.collections = [...senderAccount.collection.collections, key];
      await collectionAccountSubStore.set(context, transaction.senderAddress, senderAccount);
    } else {
      await collectionAccountSubStore.set(context, context.transaction.senderAddress, {
        collection: {
          collections: [key],
        },
      });
    }

    // Store the collection object in the blockchain
    await collectionSubStore.set(context, key, audioObject);
  }
}
