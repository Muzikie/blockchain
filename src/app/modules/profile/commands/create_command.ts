/* eslint-disable class-methods-use-this */

import {
  BaseCommand,
  CommandVerifyContext,
  CommandExecuteContext,
  VerificationResult,
  VerifyStatus,
} from 'lisk-sdk';
import { ProfileStore } from '../stores/profile';
import { ProfileAccountStore } from '../stores/profileAccount';
import { createCommandParamsSchema } from '../schemas';
import { CreateCommandParams, Profile } from '../types';
import { getNodeForName, verifyHash } from '../../../utils';

export class CreateCommand extends BaseCommand {
  public schema = createCommandParamsSchema;

  // eslint-disable-next-line @typescript-eslint/require-await
  public async verify(context: CommandVerifyContext<CreateCommandParams>): Promise<VerificationResult> {
    const { params, transaction } = context;
    // The name and nick name should not be the same
    if (params.name === params.nickName) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error('The name and nick name should not be the same'),
      };
    }
    // Validate the signature of the avatar hash
    const isAvatarHashGenuine = verifyHash(params.avatarSignature, params.avatarHash, transaction.senderPublicKey);
    if (!isAvatarHashGenuine) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error('The avatar signature is not generated by the sender'),
      };
    }
    // Validate the signature of the banner hash
    const isBannerHashGenuine = verifyHash(params.bannerSignature, params.bannerHash, transaction.senderPublicKey);
    if (!isBannerHashGenuine) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error('The banner signature is not generated by the sender'),
      };
    }
    return { status: VerifyStatus.OK };
  }

  public async execute(context: CommandExecuteContext<CreateCommandParams>): Promise<void> {
    const { params, transaction } = context;

    const profileAccountStore = this.stores.get(ProfileAccountStore);
    const profileStore = this.stores.get(ProfileStore);

    // The name should not be already registered
    const profileAccountExists = await profileAccountStore.has(context, transaction.senderAddress);
    if (profileAccountExists) {
      throw new Error('You have already created a profile for this account.');
    }

    // Get namehash output of the profile NFT
    const profileID = getNodeForName(context.transaction);

    // Create the profile account object
    const profileObject: Profile = {
      ...params,
      creatorAddress: transaction.senderAddress,
    };
    // Save the profile account object in account store
    await profileAccountStore.set(context, transaction.senderAddress, {
      profileID,
    });
    // Save the profile object in profile store
    await profileStore.set(context, profileID, profileObject);
  }
}
