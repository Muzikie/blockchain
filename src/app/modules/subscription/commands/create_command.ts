/* eslint-disable class-methods-use-this */
import {
  BaseCommand,
  CommandVerifyContext,
  CommandExecuteContext,
  VerificationResult,
  VerifyStatus,
} from 'lisk-sdk';
import { SubscriptionStore } from '../stores/subscription';
import { SubscriptionAccountStore } from '../stores/subscriptionAccount';
import { CreateCommandParams, Subscription, SubscriptionAccount } from '../types';
import { createCommandParamsSchema } from '../schemas';
import { DEV_ADDRESS } from '../constants';

export class CreateCommand extends BaseCommand {
  public schema = createCommandParamsSchema;

  // eslint-disable-next-line @typescript-eslint/require-await
  public async verify(context: CommandVerifyContext<CreateCommandParams>): Promise<VerificationResult> {
    if (!context.transaction.senderAddress.equals(DEV_ADDRESS)) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error('You are not authorized to create a subscription.'),
      }
    }
    if (context.params.maxMembers === 0 || context.params.maxMembers > 5) {
      return {
        status: VerifyStatus.FAIL,
        error: new Error('Max members of the subscription should be a number between 1 to 5.'),
      }
    }
    return { status: VerifyStatus.OK };
  }

  public async execute(context: CommandExecuteContext<CreateCommandParams>): Promise<void> {
    const {
      params: { price, maxMembers },
      transaction: { nonce, senderAddress },
    } = context;
    const subscriptionAccountStore = this.stores.get(SubscriptionAccountStore);
    const subscriptionStore = this.stores.get(SubscriptionStore);

    // Create subscription ID
    const id = Buffer.from(`subscription-${nonce}${maxMembers}${price}`, 'utf8');
    // Create subscription object
    const subscription: Subscription = {
      price,
      consumable: price, // Define the consumable equal to 80% of the price
      streams: BigInt(0),
      members: [],
      maxMembers,
      creatorAddress: senderAddress,
    };

    // Store the subscription object in the blockchain
    await subscriptionStore.set(context, id, subscription);

    // Get the sender account from the blockchain
    const senderExists = await subscriptionAccountStore.has(context, senderAddress);
    let senderAccount: SubscriptionAccount;
    if (!senderExists) {
      senderAccount = {
        subscription: {
          owned: [id],
          shared: Buffer.alloc(0),
        },
      };
    } else {
      const retrievedAccount = await subscriptionAccountStore.get(context, senderAddress);
      senderAccount = {
        subscription: {
          owned: [...retrievedAccount.subscription.owned, id],
          shared: retrievedAccount.subscription.shared,
        },
      };
    }

    // Store the account object in the blockchain
    await subscriptionAccountStore.set(context, senderAddress, senderAccount);
  }
}
