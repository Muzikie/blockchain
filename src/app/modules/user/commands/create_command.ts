/* eslint-disable class-methods-use-this */

import {
  BaseCommand,
  CommandVerifyContext,
  CommandExecuteContext,
  VerificationResult,
  VerifyStatus,
} from 'lisk-sdk';

interface Params {}

export class CreateCommand extends BaseCommand {
  public schema = {
    $id: 'CreateCommand',
    type: 'object',
    properties: {},
  };

  // eslint-disable-next-line @typescript-eslint/require-await
  public async verify(_context: CommandVerifyContext<Params>): Promise<VerificationResult> {
    return { status: VerifyStatus.OK };
  }

  public async execute(_context: CommandExecuteContext<Params>): Promise<void> {}
}
