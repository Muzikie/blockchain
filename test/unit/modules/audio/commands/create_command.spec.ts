import { CreateCommand } from '../../../../../src/app/modules/audio/commands/create_command';

describe('CreateCommand', () => {
  let command: CreateCommand;

	beforeEach(() => {
		// command = new CreateCommand();
	});

	describe('constructor', () => {
		it('should have valid name', () => {
			expect(command.name).toEqual('create');
		});

		it('should have valid schema', () => {
			expect(command.schema).toMatchSnapshot();
		});
	});

	describe('verify', () => {
		describe('schema validation', () => {
      it.todo('should throw errors for invalid schema');
      it.todo('should be ok for valid schema');
    });
	});

	describe('execute', () => {
    describe('valid cases', () => {
      it.todo('should update the state store');
    });

    describe('invalid cases', () => {
      it.todo('should throw error');
    });
	});
});
