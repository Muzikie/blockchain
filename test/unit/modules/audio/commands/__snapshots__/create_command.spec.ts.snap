// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateCommand constructor should have valid schema 1`] = `
{
  "$id": "audio/create",
  "properties": {
    "audioHash": {
      "dataType": "bytes",
      "fieldNumber": 7,
    },
    "audioSignature": {
      "dataType": "bytes",
      "fieldNumber": 6,
    },
    "collectionID": {
      "dataType": "bytes",
      "fieldNumber": 4,
    },
    "feat": {
      "fieldNumber": 8,
      "items": {
        "dataType": "bytes",
        "format": "lisk32",
      },
      "type": "array",
    },
    "genre": {
      "fieldNumber": 3,
      "items": {
        "dataType": "uint32",
      },
      "type": "array",
    },
    "name": {
      "dataType": "string",
      "fieldNumber": 1,
      "maxLength": 40,
      "minLength": 3,
    },
    "owners": {
      "fieldNumber": 5,
      "items": {
        "$id": "audio/create/owners",
        "properties": {
          "address": {
            "dataType": "bytes",
            "fieldNumber": 1,
            "format": "lisk32",
          },
          "shares": {
            "dataType": "uint32",
            "fieldNumber": 2,
          },
        },
        "required": [
          "address",
          "shares",
        ],
        "type": "object",
      },
      "type": "array",
    },
    "releaseYear": {
      "dataType": "string",
      "fieldNumber": 2,
    },
  },
  "required": [
    "name",
    "releaseYear",
    "genre",
    "collectionID",
    "owners",
    "audioSignature",
    "audioHash",
    "feat",
  ],
  "title": "CreateAsset transaction asset for audio module",
  "type": "object",
}
`;
